File: python/level2/2_spy.py
#위장
#https://programmers.co.kr/learn/courses/30/lessons/42578

import itertools


def solution(clothes):
    answer = 0
    clo_dict = {}
    for i in clothes:
        if (i[1] not in clo_dict):
            clo_dict[i[1]] = [i[0]]
        else:
            clo_dict[i[1]].append(i[0])
        answer += 1

    print(clo_dict)
    # print(answer)
    if (len(clo_dict) == 30):
        return 1073741823
    elif (len(clo_dict) > 1):
        arr = []
        p = 0
        clo_list = []
        for i in clo_dict:
            arr.append(p)
            p += 1
            clo_list.append(clo_dict[i])
        # print(clo_list)
        for i in range(1, len(clo_dict)):
            combi = list(itertools.combinations(arr, i + 1))
            # print(combi)
            for a in range(len(combi)):
                sum = 1
                for b in combi[a]:
                    sum *= len(clo_list[b])
                    # print(sum)
                answer += sum
    return answer

File: java/level2/2_spy.java
//¿ß¿Â
//https://programmers.co.kr/learn/courses/30/lessons/42578?language=java
import java.util.*;
class Solution {
    public int solution(String[][] clothes) {
        int answer = 1;
        Arrays.sort(clothes, new Comparator<String[]>() {
           public int compare(String[] o1, String[] o2) {
               return o2[1].compareTo(o1[1]);
           } 
        });
        // for (String[] i : clothes) {
        //     for (String j: i) {
        //         System.out.print(j);
        //     }
        //     System.out.println();
        // }
        ArrayList<Integer> array = new ArrayList<>();
        String now = clothes[0][1];
        int cnt = 0;
        for (int i=0; i<clothes.length; i++) {
            if (clothes[i][1].equals(now)) {
                cnt++;
            }
            else {
                array.add(cnt);
                now = clothes[i][1];
                cnt = 1;
            }
            
        }
        array.add(cnt);
        for (int i : array) {
            //System.out.print(i + " ");
            answer *= (i+1);
        }
        //System.out.println();
        
        return --answer;
    }
}